File: 208-implement-trie-prefix-tree/ken/208.py
class Trie(object):
    class Node(object):
        def __init__(self):
            self.val = False
            self.next = {}

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.node = self.Node()

    def insert(self, word):
        """
        Inserts a word into the trie.
        :type word: str
        :rtype: void
        """
        node = self.node
        for c in word:
            if c not in node.next:
                node.next[c] = self.Node()

            node = node.next[c]

        node.val = True

    def search(self, word):
        """
        Returns if the word is in the trie.
        :type word: str
        :rtype: bool
        """
        hasValue = True
        node = self.node
        for c in word:
            if c not in node.next:
                hasValue = False
                break

            node = node.next[c]

        if hasValue and node.val:
            return True
        return False

    def startsWith(self, prefix):
        """
        Returns if there is any word in the trie that starts with the given prefix.
        :type prefix: str
        :rtype: bool
        """
        hasValue = True
        node = self.node
        for c in prefix:
            if c not in node.next:
                hasValue = False
                break

            node = node.next[c]

        return hasValue

File: 208-implement-trie-prefix-tree/leochang/208.cpp
class TrieNode {
public:
    TrieNode() : isWord(false){
        next = vector<TrieNode*>(26, NULL);
    }
    bool isWord;
    vector<TrieNode*> next;
};

class Trie {
public:
    /** Initialize your data structure here. */
    Trie() {
        root = new TrieNode();
    }
    
    /** Inserts a word into the trie. */
    void insert(string word) {
        TrieNode* node = root;
        for(auto c : word){
            if(!node->next[c-'a']) {
                node->next[c-'a'] = new TrieNode();
            }
            node = node->next[c -'a'];
        }
        node->isWord = true;
    }
    
    /** Returns if the word is in the trie. */
    bool search(string word) {
        TrieNode* node = root;
        for(auto c : word) {
            if(!node->next[c-'a']) return false;
            node = node->next[c-'a'];
        }
        return node->isWord;
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    bool startsWith(string prefix) {
        TrieNode* node = root;
        for(auto c : prefix) {
            if(!node->next[c-'a']) return false;
            node = node->next[c-'a'];
        }
        return true;
    }
private:
    TrieNode* root;
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * bool param_2 = obj.search(word);
 * bool param_3 = obj.startsWith(prefix);
 */
